labs/203-custom-userspace-instrumentation-in-c/README.md

```diff
@@ -0,0 +1,283 @@
+     printf("Process %d received token %d from process %d\n", world_rank, token,
+            world_size - 1);
+   }
+```
+
+- - -
+
+### Task 4: Compile The Application
+
+The next step is to build the application with the tracepoints, using the `lttng-ust` libraries. There are many ways to compile and link an application with tracepoints: tracepoints can be statically linked with the instrumented application, or tracepoints can be compiled in a shared object, then linked with the application, or preloaded at compile time, etc. The LTTng documentation describes an [exhaustive list of compile and link scenarios](https://lttng.org/docs/v2.10/#doc-building-tracepoint-providers-and-user-application) for instrumented application.
+
+For this lab, we'll present the simplest one: statically linking the tracepoints with the application.
+
+First, we should compile the tracepoint provider package source file:
+
+```bash
+$ gcc -I. -c ring_tp.c
+```
+
+Then we compile the `ring.c` file with the tracepoint file and the lttng-ust library flag.
+
+```bash
+$ mpicc -o ring ring_tp.o ring.c -llttng-ust -ldl
+```
+
+The `ring` application is now instrumented with static LTTng tracepoints. The application can be run as usual and the tracepoints, when not traced, should have a near zero overhead.
+
+```bash
+$ mpirun -N 4 ./ring
+Process 1 received token -1 from process 0
+Process 2 received token -1 from process 1
+Process 3 received token -1 from process 2
+Process 0 received token -1 from process 3
+```
+
+The following block shows the difference between the [original simple `makefile`](code/makefile.orig) and the [`makefile` for static linkage](code/makefile) of tracepoint code.
+
+```diff
+@@ -4,7 +4,8 @@ MPICC?=mpicc
+ all: ${EXECS}
+
+ ring: ring.c
+-       ${MPICC} -o ring ring.c
++       gcc -I. -c ring_tp.c
++       ${MPICC} -o ring ring_tp.o ring.c -llttng-ust -ldl
+
+ clean:
+        rm -f ${EXECS}
+```
+
+- - -
+
+### Task 5: Trace an Instrumented Application
+
+To trace an application instrumented with LTTng-UST tracepoints, one simply needs to create a tracing session, enable the userspace events corresponding to the **ring** tracepoint provider, start tracing and run the application.
+
+Since we are tracing userspace only, we add the `vtid` context, to differentiate the running thread. This context wouldn't be necessary if there was a kernel trace taken at the same time.
+
+```bash
+$ lttng create
+Session auto-20191015-113432 created.
+Traces will be output to /home/user/lttng-traces/auto-20191015-113432
+$ lttng enable-event -u 'ring:*'
+UST event ring:* created in channel channel0
+$ lttng add-context -u -t vtid
+UST context vtid added to all channels
+$ lttng start
+Tracing started for session auto-20191015-113432
+$ mpirun -N 4 ./code/ring
+Process 1 received token -1 from process 0
+Process 2 received token -1 from process 1
+Process 3 received token -1 from process 2
+Process 0 received token -1 from process 3
+$ lttng destroy
```

Comment:
You could do 'lttng stop', 'lttng view' to make sure some events were recorded and then 'lttng destroy'. It will have the added bonus of printing the full path to the trace directory for use in further steps.